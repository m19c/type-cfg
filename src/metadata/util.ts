import { ReflectionMissingError } from '../errors/ReflectionMissingError';

export function ensureReflectMetadataExists(): void {
  if (
    typeof Reflect !== 'object' ||
    typeof Reflect.decorate !== 'function' ||
    typeof Reflect.metadata !== 'function'
  ) {
    throw new ReflectionMissingError();
  }
}

export function getGlobalVariable(): typeof global {
  return global;
}
