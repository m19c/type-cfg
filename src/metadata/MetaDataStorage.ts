import { ensureReflectMetadataExists } from './util';
import { DefinitionMetadata } from './definitions/DefinitionMetadata';
import { PropertyMetadata } from './definitions/PropertyMetadata';
import { EnumMetadata } from './definitions/EnumMetadata';

export class MetadataStorage {
  private definitions: DefinitionMetadata[] = [];
  private properties: PropertyMetadata[] = [];
  private enums: EnumMetadata[] = [];

  constructor() {
    ensureReflectMetadataExists();
  }

  collectDefinitionMetadata(item: DefinitionMetadata) {
    this.definitions.push(item);
  }

  findDefinitionByInstance(value: any) {
    return (
      this.definitions.filter(definition => value instanceof definition.target).shift() || null
    );
  }

  findDefinitionByValue(value: any) {
    return this.definitions.filter(definition => value === definition.target).shift() || null;
  }

  collectConfigPropertyMetadata(item: PropertyMetadata) {
    this.properties.push(item);
  }

  findPropertiesByInstance(value: any) {
    return this.properties.filter(property => value instanceof property.target);
  }

  collectEnumMetadata(item: EnumMetadata) {
    this.enums.push(item);
  }

  findEnum(value: any) {
    return this.enums.filter(item => item.value === value).shift() || null;
  }
}
