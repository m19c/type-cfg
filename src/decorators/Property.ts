import { TypeFunction } from './types';
import { getMetadataStorage } from '../metadata';
import { createTypeDeterminerFunction } from '../util/type';
import { resolveTypeDecoratorArguments } from '../util/arg';
import { PropertyMetadataDefaultValue } from '../metadata/definitions/PropertyMetadata';

export interface PropertyOptions {
  source?: string;
  delimiter?: string;
  required?: boolean;
  defaultValue?: PropertyMetadataDefaultValue;
}

// eslint-disable-next-line import/export
export function Property(): PropertyDecorator;
// eslint-disable-next-line import/export
export function Property(options: PropertyOptions): PropertyDecorator;
// eslint-disable-next-line import/export
export function Property(typeFunction: TypeFunction): PropertyDecorator;
// eslint-disable-next-line import/export
export function Property(typeFunction: TypeFunction, options: PropertyOptions): PropertyDecorator;

// eslint-disable-next-line import/export
export function Property(
  typeFunctionOrOptions?: TypeFunction | PropertyOptions,
  maybeOptions?: PropertyOptions
): PropertyDecorator {
  return function captureConfigPropertyTarget(prototype, propertyKey): void {
    const name = propertyKey.toString();
    const { options, typeFunction } = resolveTypeDecoratorArguments(
      typeFunctionOrOptions,
      maybeOptions
    );

    getMetadataStorage().collectConfigPropertyMetadata({
      name,
      target: prototype.constructor,
      ...createTypeDeterminerFunction({
        prototype,
        propertyKey: name,
        typeFunction,
      }),
      source: options.source || null,
      delimiter: options.delimiter || null,
      required: options.required || true,
      defaultValue: options.defaultValue,
    });
  };
}
