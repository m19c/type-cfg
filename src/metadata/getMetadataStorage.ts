import { MetadataStorage } from './MetaDataStorage';
import { getGlobalVariable } from './util';

declare global {
  // eslint-disable-next-line @typescript-eslint/no-namespace
  namespace NodeJS {
    interface Global {
      TypeConfigMetadataStorage?: MetadataStorage;
    }
  }
}

export function getMetadataStorage(): MetadataStorage {
  const root = getGlobalVariable();

  if (!root.TypeConfigMetadataStorage) {
    root.TypeConfigMetadataStorage = new MetadataStorage();
  }

  return root.TypeConfigMetadataStorage;
}
