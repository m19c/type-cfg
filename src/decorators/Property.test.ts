import * as md from '../metadata';
import { Property } from './Property';

describe('Property', () => {
  test('calls the metadata storage collector', () => {
    const storage = {
      collectPropertyMetadata: jest.fn(),
    };
    // eslint-disable-next-line no-multi-assign
    const getMetadataStorage = ((md as any).getMetadataStorage = jest.fn(() => storage));

    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    class TargetClass {
      @Property()
      test: string;
    }

    expect(storage.collectPropertyMetadata).toHaveBeenCalled();
    expect(getMetadataStorage).toHaveBeenCalled();

    const [first] = storage.collectPropertyMetadata.mock.calls;
    expect(first).toMatchSnapshot();
  });
});
