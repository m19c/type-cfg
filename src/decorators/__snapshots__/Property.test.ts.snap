// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Property calls the metadata storage collector 1`] = `
Array [
  Object {
    "defaultValue": undefined,
    "delimiter": null,
    "determineType": [Function],
    "determineTypeOptions": Object {
      "isArray": false,
    },
    "name": "test",
    "required": true,
    "source": null,
    "target": [Function],
  },
]
`;
